File: dsa-py/queue/circular-queue.py
SIZE = 5
queue = [None]*SIZE
front = -1
rear = -1

def isEmpty():
    return front == -1

def isFull():
    return (rear+1)%SIZE == front

def display():
    if isEmpty():
        print("Queue empty")
        return
    i = front
    while True:
        print(queue[i], end=" ")
        if i == rear:
            break
        i = (i+1)%SIZE
    print()

def enqueue(value):
    global front, rear
    if isFull(): return
    if(front == -1):
        front = 0
    rear = (rear+1)%SIZE
    queue[rear] = value   
    print("Enqueue: ", value)

def dequeue():
    global front, rear
    if isEmpty(): 
        print("empty, cannot remove")
        return
    data = queue[front]
    print("Dequeue: ", data)
    if front==rear:
        front = rear = -1
    else:
        front = (front+1)%SIZE

display()
enqueue(10)
enqueue(20)
enqueue(30)
enqueue(40)
enqueue(50)
enqueue(60)
display()
dequeue()
display()
dequeue()
display()
dequeue() 
dequeue()
display()
enqueue(14)
display()

File: dsa-c/queue-c/circular-queue.c
#include<stdio.h>
#define MAX 5

struct CQueue{
    int items[MAX];
    int front;
    int rear;
};
typedef struct CQueue CQueue;

void initQueue(CQueue* c){
   c->front = c->rear = -1;
}

int isEmpty(CQueue* c){
    return c->front == -1;
}
int isFull(CQueue* c){
    return (c->rear+1)%MAX == c->front;
}
void enqueue(CQueue* c, int value){
    if(isFull(c)){printf("Queue is full \n");}
    else{
        if(isEmpty(c)){c->front = c->rear = 0;}
        else{
            c->rear = (c->rear+1)%MAX;
        }
        c->items[c->rear] = value;
    }
}

void dequeue(CQueue* c){
    if(isEmpty(c)){printf("Queue underflow \n");}
    else{
        int data = c->items[c->front];
        if(c->front == c->rear){c->front = c->rear = -1;}
        else{
            c->front = (c->front + 1)%MAX;
        }
        printf("Dequeued: %d \n", data);
    }
}

void display(CQueue* c){
    printf("Display queue \n");
    int i = c->front;
    while(1){
        printf("%d \n",c->items[i]);
        if(i == c->rear){break;}
        i =(i+1)%MAX;
    }
}

int main(){
    CQueue c;
    initQueue(&c);
    enqueue(&c,2), enqueue(&c,7), enqueue(&c,3),enqueue(&c,6), enqueue(&c,10);
    display(&c);
    enqueue(&c,13);
    dequeue(&c);
    display(&c);
    enqueue(&c,13);
    display(&c);

    return 0;
}
